### DISCUSSION TITLE
Expose a method to retrieve a user by email address

### DISCUSSION THREAD

**Discussion by yagiz-noonlight** [supabase/supabase#13459] [Feature Requests]
Currently Supabase does not support retrieving a user other than `uid`. For applications that involve an invitation flow, where the developer needs to check if that email belongs to an existing user or not, there is currently no way to know it, other than, getting all users and filtering them through email on the application layer.

Something like this:

`supabase.auth.admin.getUserByEmail()`

**Comment by tazmaniax** [comment]
I guess the workaround for this would be to create a [user profile table in the public schema](https://supabase.com/docs/guides/auth/managing-user-data#creating-user-tables) (e.g. `public.users`) that is auto-populated when a record is added to the `auth.users` table. You would need to modify the trigger on `auth.users`, see [Advanced techniques](https://supabase.com/docs/guides/auth/managing-user-data#advanced-techniques), to copy over the email address, e.g.
```PLpgSQL
-- inserts a row into public.users
create function public.handle_new_user()
returns trigger
language plpgsql
security definer set search_path = public
as $$
begin
  insert into public.users (id, email)
  values (new.id, new.email);
  return new;
end;
$$;
```
As `public.users` is assured to be in sync with `auth.users` using the insert trigger and the cascade delete you can reliably search `public.users` by email address. 
Creating a user profile table is probably the common case so it would be nice to see this setup automated, however I was able to set this all up relatively quickly using the Supabase console (database functions & triggers). When you add the insert function through the console you only need the following
```PLpgSQL
begin
  insert into public.users (id, email)
  values (new.id, new.email);
  return new;
end;
```

I also created an update function and trigger for good measure in case the email could be modified.
```PLpgSQL
begin
  update public.users
  set email = new.email
  where id = new.id;
  return new;
end;
```
Just make sure you set the "Type of security" to "SECURITY DEFINER" and for good measure, I added the config param "search_path" = "public", although I wasn't 100% sure this was required.
![CleanShot 2023-04-09 at 18 59 59@2x](https://user-images.githubusercontent.com/1079943/230788919-7aae72e3-6859-4d8c-b66c-4a5122fcf6d1.png)

This approach also means you can't have any mandatory fields (e.g. first_name, last_name) apart from id and email which are guaranteed.

A couple of references on the topic:
* [Supabase: how to query users by eMail?](https://stackoverflow.com/questions/68334303/supabase-how-to-query-users-by-email)
* [Supabase: How to query users table?](https://nikofischer.com/supabase-how-to-query-users-table)
